GVFS/GVFS.Common/FileBasedCollections/BinaryPlaceholderListDatabase.cs

```diff
@@ -0,0 +1,358 @@
+﻿using GVFS.Common.FileBasedCollections;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace GVFS.Common.FileBasedCollections
+{
+    public class BinaryPlaceholderListDatabase : BinaryFileBasedCollection<PlaceholderEvent>

```

Comment:
I compared this file to `PlaceholderListDatabase.cs` and found the changes mostly related to the new type structure, with a little bit of logic around the "remove" events.

While the changes looked appropriate, I can't shake a feeling that this "copy and update types" strategy is smelly and likely to include bugs. However, sometimes we need a big change because the old code is old and poorly typed.

The best thing I could think to do is to create an interface between the placeholder list database logic and the file based collection logic, and that can use the new type structure around `PlaceholderEvent`. Then, create an adapter class that "upgrades" the old `FileBasedCollection` to create the new types. Then, we don't need to have two classes that are mostly duplicates here, and don't need to `[Obsolete]` the other class.

One obstruction I see to this plan is that we have some coupling in this class to the underlying file format. That's already something I'd like to remedy, and that would make the classes more independent. In general, I prefer to have only one class know about a file format, and everything else be independent.